### ISSUE TITLE
within=caliper with NA's

### ISSUE THREAD

**Issue by josherrickson** [markmfredrickson/optmatch#147]
### remaining to dos on this issue

- review associated branch
- update as needed
- merge into master
- update release notes

### issue description

This may be related to #142 but seems distinct.

```
> nuclearplants$cost[2] <- NA
> c <- caliper(match_on(pr ~ cost, data = nuclearplants, method = "euclidean"), width = 100)
> match_on(pr ~ cost, within = c, data = nuclearplants)
 Error in makedist(z, data, compute_mahalanobis, within) : 
  Row and column names of within must match those of the data. 
```

I believe this is due to `cost` being both involved in creating the caliper and also using to create the distance. This does not error without `~ cost` in `match_on`, e.g.

```
> m <- match_on(pr ~ cap, within = c, data = nuclearplants)
> dim(m)
[1] 10 22
```

We can get around this by running `match_on` without `within` and instead adding the caliper to it. 

```
> np <- nuclearplants[1:6,]
> c <- caliper(match_on(pr ~ cost, data = np, method = "euclidean"), width = 100)
> match_on(pr ~ cap, within = c, data = np)
       control
treated        H   J        K   I
      A 1.573891 Inf 2.294217 Inf
      B      Inf   0      Inf Inf
> m <- match_on(pr ~ cap, data = np)
> m + c
       control
treated        H   I   J        K
      A 1.573891 Inf Inf 2.294217
      B      Inf Inf   0      Inf
```

**Comment by benthestatistician** [comment]
The problem also has to something do with the NA on the RHS of the formula given to `match_on()`. 

```{r}
> dim( match_on(pr ~ cap, within = c, data = nuclearplants) )
 [1] 10 22
> nuclearplants$cap[2] <- NA
> dim( match_on(pr ~ cap, within = c, data = nuclearplants) )
Error in makedist(z, data, compute_mahalanobis, within) : 
  Row and column names of within must match those of the data.
```

**Comment by benthestatistician** [comment]
Started a new branch for this problem, issue147.

**Comment by adamrauh** [comment]
My findings are consistent with @benthestatistician so far. I added in a subset to remove the rows in columns not appearing in mf before the call to `makedist()` in `match_on.formula()` and this seems to stop the error in the cases here. For instance, 
```r
> dim( match_on(pr ~ cap, within = c, data = nuclearplants) )
[1] 10 22
> nuclearplants$cap[2] <- NA
> dim( match_on(pr ~ cap, within = c, data = nuclearplants) )
[1] 10 22
```
I'm going to check this a little bit more and see if it passes the tests, then probably commit to the issue147 branch if things check out.

**Comment by adamrauh** [comment]
working on fixes [here](https://github.com/adamrauh/optmatch/commit/36f2633ccb0caad3955ff6c3cf81846fa7aa984a#diff-e1e1d3d40573127e9ee0480caf1283d6)

**Comment by benthestatistician** [comment]
In conversation just now it occurred to @adamrauh and me that one of my earlier comments had set him somewhat on the wrong track.  I was forgetting about some of the internal presumptions of the codebase.  His code suggestions are good but will have to be updated.  (I'm deliberately leaving this a bit cryptic; Adam do add updates and clarifications as appropriate.)

Underlying this: for `match_on()` to omit and stop tracking observations because of NAs is the wrong answer.  It might refuse to match those observations to anyone else, but their identities should be retained, in the rownames and/or colnames slot.  (At least, that's my current understanding of how it ought to be: do speak up @josherrickson if you see it differently.)

**Comment by adamrauh** [comment]
The general idea of my current thoughts are to perform a similar, but slightly different set of operations on the data in the `makedist()` function. One thing @benthestatistician brought up, which I hadn't considered, is avoiding making extra copies of objects (especially `within`) wherever possible, and I agree that moving this kind of code to `makedist()` would help. There is also some logic at the end of `match_on.formula()` that re-adds dropped units and other related information. I think doing some intelligent subsetting in the appropriate places within `makedist()` would let us take out some of that kind of code in `match_on.formula()`

I think I can use the conditions that are causing the error right now to help with this subsetting. Based on what Ben said, it sounds like the reasoning behind the conditions: `!all(within@rownames %in% rns)` and `!all(within@colnames %in% cns)` is a bit unclear (in makedist.R, lines 65-68). If this is the case, these expressions would be useful to help with subsetting the data. I think updates to the rows, cols, and .data slots would be required, but I think (at this point, anyway) that the other slots might not be affected. Then, I don't think other code would be necessary to re-add any information. I'm a bit hesitant to remove those conditions entirely, but I'll see what happens.

**Comment by benthestatistician** [comment]
Thanks for posting this update, @adamrauh .  I agree that [at L65-L66 of makedist.R](https://github.com/markmfredrickson/optmatch/blob/master/R/makedist.R#L65)  we should strip out the conditions `!all(within@rownames %in% runs)` and `!all(within@colnames %in% cos)`.

**Comment by adamrauh** [comment]
I just pushed up some new changes, which you can see in [this commit](https://github.com/adamrauh/optmatch/commit/074cc0644a7d924f44b02aafb69d824cb0386fdc?w=1). I basically moved where the subsetting happens from `match_on.formula()` to `makedist()` and then ended up adding a few patches to make sure that information about dropped units gets added back at the end. I'm not entirely satisfied with this solution because it still relies on subsetting + re-adding. I've been trying to figure out a way to get the desired result without having to keep track of dropped unit info and re-adding it, but all my solutions thus far end up causing a lot of problems elsewhere in the code.

**Comment by benthestatistician** [comment]
I like the direction of @adamrauh 's commit.   

Before commenting more, I'd like some feedback from @markmfredrickson, or @josherrickson, on a fine point about outmatch objects' interactions with one another.   Would you agree with the following.

1.  Suppose that `match_on` receives both a within argument and a data argument. If the within object references subjects not represented as rows in the data argument, my feeling is that those subjects should not be represented in what `match_on` returns.  
2. If there are subjects appearing as rows of the data argument that aren't mentioned as row or column names of a received within argument, it should be an error. 
3.  To summarize, for `match_on`'s purposes, the "sample", so to speak, is defined by the `data` argument.

**Comment by benthestatistician** [comment]
@adamrauh once we're all agreed on what the expectations ought to be vis a vis compatibility of names provided via data and within argument, we should write tests to enforce these expectations, committing them in a dedicated commit that references this issue thread.  I'll leave this in your court.

**Comment by adamrauh** [comment]
@benthestatistician Noted. Based on one of my non-functional attempts at a solution yesterday, it appears that there might be some assumptions about ISMs/BISMs and `within` objects that aren't outwardly obvious (to me, anyway). As I mentioned, I ran into some issues elsewhere in the code when I attempted to subset the rows, cols, and .Data slots (and not rownames/colnames) with the intent of avoiding keeping track of dropped information, so I'm curious to learn more here. It could have been just me not performing the operations correctly, but I lean towards thinking I violated some assumptions along the way.

**Comment by josherrickson** [comment]
I had a thought of a different approach to this (it wouldn't completely remove the need to address this [see below] but does remove the most common cause).

It seems to me that the primary culprit here is that an NA on the RHS of a `caliper` or `exactMatch` results in that column/row being dropped completely. Is this the most expected outcome? Would it not make more sense include the NA'd row/column, and have it Inf'd out against everything?

That would fix the errors that opened this issue. We'd still need to come up with a solution in case people generated their own mask, but I think we could be more strict then, forcing it to have the exact same rows/columns as `data`.

**Comment by adamrauh** [comment]
@josherrickson I believe I actually experimented with the first situation you describe-- with the NA'd unit not being dropped, but included with Infs. I think it caused problems in either the `issue.warning <- issue.warning || (dim(prepareMatching(s))[1] > getMaxProblemSize())` step or at `dists <- distancefn(cbind(treatmentids, controlids), data, z)` in `makedist.R`... could be my fault, though. I can revisit this and post some results.

**Comment by josherrickson** [comment]
@adamrauh I imagine this might require quite a bit more restructuring then the current proposal. But it might smooth the way in the future. I'm curious what @benthestatistician or @markmfredrickson think.

**Comment by adamrauh** [comment]
Within `makedist.R` I tried:

```{r}
within@rows <- within@rows[within@cols %in% which(within@colnames %in% cns)]
within@cols <- within@cols[within@cols %in% which(within@colnames %in% cns)]
within@.Data <- within@.Data[within@cols %in% which(within@colnames %in% cns)]
```
which seems to give:
```{r}
$`0`
       control
treated H   I J K L M N O P Q R S T U V W X Y Z
      A 0 Inf 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
      B 0 Inf 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
      C 0 Inf 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
      D 0 Inf 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
      E 0 Inf 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
      F 0 Inf 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
      G 0 Inf 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0

$`1`
       control
treated d e f
      a 0 0 0
      b 0 0 0
      c 0 0 0
```
and then:
```{r}
subprobs <- findSubproblems(within)
for (s in subprobs) {
      issue.warning <- issue.warning || (dim(prepareMatching(s))[1] > getMaxProblemSize())
    }
```
returns
```{r}
Error in data.frame(control = as.factor(distances@colnames[distances@cols]),  : 
  arguments imply differing number of rows: 126, 133
```
which is why I reverted back to the other idea.

**Comment by benthestatistician** [comment]
My reactions to recent comments of @adamrauh and @josherrickson, in reverse-importance order:

1. Re Josh's thought about new caliper argument (`na.action = {Inf, 0}`): I'd be somewhat averse to adding a new `caliper()` argument for this purpose, particularly since something closely related is already available using the `exclude` argument.    On the other hand I agree that it makes sense to support the use cases Josh has in mind.  I'd think the first steps in that direction would be to construct some examples of those use cases.  Perhaps if we're clever enough we can handle them simply enough using `exclude=`, showing others how to do it with examples in the man page. If that turns out to be difficult then we might think about making it easier by writing a helper function, `optmatch::na_omits()` or something, accepting a data frame or fitted model and returning a character vector of row names of incomplete cases, as expected in the `exclude=` position, and enriching the docs to include some examples demonstrating its use.
2. Re Adam's not-quite successful `makedist` update, I'd expect an error of this sort as a result of not updating `within@groups` in parallel w/ updates to other slots. (That may or may not ID the issue; perhaps it's a hint.) 
3. I think I'd agree w/ Josh's take on how NAs ought to be handled by `match_on()` and related functions.  But I'd prefer to think more and perhaps discuss before having that work comments.  I'm more sure that I agree that it's a project w/ at least somewhat wider scope.  
4. If I'm not mistaken, Adam's close to having this issue licked as is.  If no one disagrees with that, then I think we should encourage him to see his solution through before considering potential additional benefits of more work along these lines.  (We are approaching a deadline for keeping the package on CRAN!)
4. Speaking of disagreements, I hear Josh speaking up for the idea of insisting that the `within` argument and the `data` argument observe precisely the same row names, a premise I've been arguing for getting away from, in this thread and in #142.   (I think that `data`'s row names should have to be among those recorded in the `within`, but not conversely. )  Speaking of approaching deadlines, I'd like to invite a f2f discussion at a later time of merits of that premise, while in the immediate term relaxing it in the manner Adam does [here](https://github.com/adamrauh/optmatch/blob/issue147/R/makedist.R#L69).

**Comment by adamrauh** [comment]
On @benthestatistician 's point 2: It looks like updating within@groups removes the corresponding row/col from the matrix. Plus, the same error still persists.

and on 4: the issue147 branch on my fork should pass all existing tests and solves the issue (as I understand it). I think adding some additional tests would be a good idea, in light of some of the discoveries.

**Comment by josherrickson** [comment]
Re 3/4: Sounds reasonable.

Re 5: I think you're right - within can have excess (if you subset the data for example) but should be missing anything. I agree with your comment above (https://github.com/markmfredrickson/optmatch/issues/147#issuecomment-387766320).

**Comment by adamrauh** [comment]
Just noting that if we feel okay with the approach I’ve got for now, I think it’s pretty much ready for inspection.

**Comment by benthestatistician** [comment]
thanks, good to hear!   Looking over your commits i like the jist of what you've done, but I'll wait for the pull request to view them all together.  

En route to that, do implement the full 5 steps @josherrickson & I had suggested by email.  E.g., I take it all tests are passing now in your issue147 branch?  Once that's the case, don't forget to pull in recent changes to public master, first into your forked master and then into your issue147 branch, and make sure they don't break stuff. Free free to request a pull from your issue147 branch into Mark's.  I.e., better that you leave the merge of that branch back into master to be done over here -- but after  checking that it shouldn't break stuff by merging master into it.

(There's another git process-oriented request that I just conveyed by way of a commit comment.)

**Comment by adamrauh** [comment]
@benthestatistician All the tests are passing, including some new ones that I added in. More specifically, I added in a new `test_that()` function with relevant material. I'll try to implement the overall process outlined previously to ensure that nothing breaks.

**Comment by adamrauh** [comment]
After discussion with @benthestatistician (following his discussion with @josherrickson), some updates to the documentation of `InfinitySparseMatrix.R` are needed. The plan is to update the docs, include descriptions of the different slots, and note that ordering of rows/cols may change as a result of subsetting and/or other manipulations. I think we've agreed to close this issue when that has been done sufficiently-- which I am working on.

**Comment by benthestatistician** [comment]
Nice progress.  As I noted in a comment on the commit, corresponding docs for BISMs would make sense.  Not sure about DenseMatrices; that might be in order also. 

Since @markmfredrickson designed and implemented ISMs and BISMs and the like, I'd direct the pull request to him when ready.

**Comment by benthestatistician** [comment]
I've pulled Adam's changes into the issue147 branch.  All that remains is to merge this branch into master.  (No reason not to, just leaving this for tomorrow.)

**Comment by benthestatistician** [comment]
I've brought the issue147 branch up to date with the master branch.  Passes make test, if I wasn't able to confirm make check b/c of an issue w make document that seems likely to be an idiosyncracy of my system.

**Comment by josherrickson** [comment]
Check passes, closing.
